File: tests/test_config.py
from __future__ import annotations

import os
import stat

import pytest

from lcpush import paths
from lcpush.config import (
    Config,
    from_dict,
    load,
    parse_repo,
    render_show,
    save,
    set_value,
)
from lcpush.errors import ConfigError


def test_defaults_match_spec():
    config = Config()
    assert config.commit.message_template == "Add {id}. {title} ({language})"
    assert config.commit.update_template == "Update {id}. {title} ({language})"
    assert config.commit.prompt == "confirm"
    assert config.cache.problems_ttl_days == 7
    assert config.repo.branch == "main"


@pytest.mark.parametrize(
    "value",
    [
        "user/leetcode-solutions",
        "https://github.com/user/leetcode-solutions",
        "git@github.com:user/leetcode-solutions.git",
        "  user/leetcode-solutions/  ",
    ],
)
def test_parse_repo_accepts_common_forms(value):
    assert parse_repo(value) == ("user", "leetcode-solutions")


@pytest.mark.parametrize("value", ["user", "a/b/c", ""])
def test_parse_repo_rejects_junk(value):
    with pytest.raises(ConfigError):
        parse_repo(value)


def test_round_trip(tmp_path):
    config = set_value(Config(), "repo", "user/solutions")
    path = save(config)
    assert path == paths.config_file()
    assert load() == config


def test_config_file_is_0600():
    save(Config())
    mode = stat.S_IMODE(os.stat(paths.config_file()).st_mode)
    assert mode == 0o600


def test_load_returns_none_without_a_file():
    assert load() is None


def test_load_rejects_broken_toml():
    path = paths.config_file()
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text("this is not = = toml", encoding="utf-8")
    with pytest.raises(ConfigError):
        load()


def test_set_repo_and_branch_and_path():
    config = set_value(Config(), "repo", "me/solutions")
    config = set_value(config, "branch", "trunk")
    config = set_value(config, "path", "solutions")
    assert config.repo.full_name == "me/solutions"
    assert config.repo.branch == "trunk"
    assert config.repo.path == "solutions/"


def test_path_prefix_normalizes_slashes():
    assert set_value(Config(), "path", "/a/b/").repo.path == "a/b/"
    assert set_value(Config(), "path", "  ").repo.path == ""


def test_set_is_immutable():
    original = Config()
    updated = set_value(original, "branch", "dev")
    assert original.repo.branch == "main"
    assert updated.repo.branch == "dev"


def test_set_commit_prompt_validates():
    assert set_value(Config(), "commit.prompt", "always").commit.prompt == "always"
    with pytest.raises(ConfigError):
        set_value(Config(), "commit.prompt", "sometimes")


def test_set_rejects_unknown_key():
    with pytest.raises(ConfigError):
        set_value(Config(), "repo.owner", "me")


def test_set_rejects_empty_template_and_branch():
    with pytest.raises(ConfigError):
        set_value(Config(), "commit.message_template", "   ")
    with pytest.raises(ConfigError):
        set_value(Config(), "branch", " ")


def test_set_ttl_validates():
    assert set_value(Config(), "cache.problems_ttl_days", "30").cache.problems_ttl_days == 30
    with pytest.raises(ConfigError):
        set_value(Config(), "cache.problems_ttl_days", "soon")
    with pytest.raises(ConfigError):
        set_value(Config(), "cache.problems_ttl_days", "-1")


def test_from_dict_falls_back_on_garbage():
    config = from_dict({"cache": {"problems_ttl_days": "many"}, "commit": {"prompt": "x"}})
    assert config.cache.problems_ttl_days == 7
    assert config.commit.prompt == "confirm"


def test_show_never_contains_a_token():
    config = set_value(Config(), "repo", "user/solutions")
    output = render_show(config, token_present=True, token_source="keyring")
    assert "ghp_" not in output
    assert "redacted" in output
    assert "user" in output

File: tests/test_config.cpp
#include <catch2/catch_test_macros.hpp>
#include <catch2/generators/catch_generators.hpp>

#include <sys/stat.h>

#include <fstream>

#include "helpers/fixtures.hpp"
#include "lcpush/config/config.hpp"
#include "lcpush/core/errors.hpp"
#include "lcpush/platform/paths.hpp"

using namespace lcpush;

TEST_CASE("config defaults match spec") {
    config::Config cfg;
    CHECK(cfg.commit.message_template == "Add {id}. {title} ({language})");
    CHECK(cfg.commit.update_template == "Update {id}. {title} ({language})");
    CHECK(cfg.commit.prompt == "confirm");
    CHECK(cfg.cache.problems_ttl_days == 7);
    CHECK(cfg.repo.branch == "main");
}

TEST_CASE("parse repo accepts common forms") {
    auto value = GENERATE(as<std::string>{}, "user/leetcode-solutions",
                          "https://github.com/user/leetcode-solutions",
                          "git@github.com:user/leetcode-solutions.git",
                          "  user/leetcode-solutions/  ");
    auto [owner, name] = config::parse_repo(value);
    CHECK(owner == "user");
    CHECK(name == "leetcode-solutions");
}

TEST_CASE("parse repo rejects junk") {
    auto value = GENERATE(as<std::string>{}, "user", "a/b/c", "");
    CHECK_THROWS_AS(config::parse_repo(value), ConfigError);
}

TEST_CASE("config round trip") {
    testing::IsolatedDirs dirs;
    auto cfg = config::set_value(config::Config{}, "repo", "user/solutions");
    auto path = config::save(cfg);
    CHECK(path == paths::config_file());
    auto loaded = config::load();
    REQUIRE(loaded.has_value());
    CHECK(*loaded == cfg);
}

TEST_CASE("config file is 0600") {
    testing::IsolatedDirs dirs;
    config::save(config::Config{});
    struct stat info{};
    REQUIRE(::stat(paths::config_file().c_str(), &info) == 0);
    CHECK((info.st_mode & 0777) == 0600);
}

TEST_CASE("load returns none without a file") {
    testing::IsolatedDirs dirs;
    CHECK_FALSE(config::load().has_value());
}

TEST_CASE("load rejects broken toml") {
    testing::IsolatedDirs dirs;
    auto path = paths::config_file();
    std::filesystem::create_directories(path.parent_path());
    std::ofstream(path) << "this is not = = toml";
    CHECK_THROWS_AS(config::load(), ConfigError);
}

TEST_CASE("set repo and branch and path") {
    auto cfg = config::set_value(config::Config{}, "repo", "me/solutions");
    cfg = config::set_value(cfg, "branch", "trunk");
    cfg = config::set_value(cfg, "path", "solutions");
    CHECK(cfg.repo.full_name() == "me/solutions");
    CHECK(cfg.repo.branch == "trunk");
    CHECK(cfg.repo.path == "solutions/");
}

TEST_CASE("path prefix normalizes slashes") {
    CHECK(config::set_value(config::Config{}, "path", "/a/b/").repo.path == "a/b/");
    CHECK(config::set_value(config::Config{}, "path", "  ").repo.path == "");
}

TEST_CASE("set is immutable") {
    config::Config original;
    auto updated = config::set_value(original, "branch", "dev");
    CHECK(original.repo.branch == "main");
    CHECK(updated.repo.branch == "dev");
}

TEST_CASE("set commit prompt validates") {
    CHECK(config::set_value(config::Config{}, "commit.prompt", "always").commit.prompt ==
          "always");
    CHECK_THROWS_AS(config::set_value(config::Config{}, "commit.prompt", "sometimes"),
                    ConfigError);
}

TEST_CASE("set rejects unknown key") {
    CHECK_THROWS_AS(config::set_value(config::Config{}, "repo.owner", "me"), ConfigError);
}

TEST_CASE("set rejects empty template and branch") {
    CHECK_THROWS_AS(config::set_value(config::Config{}, "commit.message_template", "   "),
                    ConfigError);
    CHECK_THROWS_AS(config::set_value(config::Config{}, "branch", " "), ConfigError);
}

TEST_CASE("set ttl validates") {
    CHECK(config::set_value(config::Config{}, "cache.problems_ttl_days", "30")
              .cache.problems_ttl_days == 30);
    CHECK_THROWS_AS(config::set_value(config::Config{}, "cache.problems_ttl_days", "soon"),
                    ConfigError);
    CHECK_THROWS_AS(config::set_value(config::Config{}, "cache.problems_ttl_days", "-1"),
                    ConfigError);
}

TEST_CASE("load falls back on garbage values") {
    testing::IsolatedDirs dirs;
    auto path = paths::config_file();
    std::filesystem::create_directories(path.parent_path());
    std::ofstream(path) << "[cache]\nproblems_ttl_days = \"many\"\n"
                        << "[commit]\nprompt = \"x\"\n";
    auto cfg = config::load();
    REQUIRE(cfg.has_value());
    CHECK(cfg->cache.problems_ttl_days == 7);
    CHECK(cfg->commit.prompt == "confirm");
}

TEST_CASE("show never contains a token") {
    auto cfg = config::set_value(config::Config{}, "repo", "user/solutions");
    auto output = config::render_show(cfg, true, "keyring");
    CHECK(output.find("ghp_") == std::string::npos);
    CHECK(output.find("redacted") != std::string::npos);
    CHECK(output.find("user") != std::string::npos);
}

TEST_CASE("dumps matches the python tomli_w layout") {
    config::Config cfg;
    cfg.repo.owner = "user";
    cfg.repo.name = "leetcode-solutions";
    std::string expected =
        "[repo]\n"
        "owner = \"user\"\n"
        "name = \"leetcode-solutions\"\n"
        "branch = \"main\"\n"
        "path = \"\"\n"
        "\n"
        "[commit]\n"
        "message_template = \"Add {id}. {title} ({language})\"\n"
        "update_template = \"Update {id}. {title} ({language})\"\n"
        "prompt = \"confirm\"\n"
        "author_name = \"\"\n"
        "author_email = \"\"\n"
        "\n"
        "[cache]\n"
        "problems_ttl_days = 7\n";
    CHECK(config::dumps(cfg) == expected);
}
